src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Connection.cs

```diff
@@ -384,5 +401,69 @@ private enum ConnectionState
             Disconnecting,
             SocketClosed
         }
+
+        private class BufferLengthControl : IBufferLengthControl
```

Comment:
Since this is effectively used by multiple classes (even if it is via an interface), I think I like this in it's own file. I think already Kestrel uses nested classes too heavily.